### PR TITLE
Optimize cache usage in DefaultConversionService and fix race condition

### LINKED ISSUE #10091: Race condition on concurrent ApplicationContext start [3.10.x]
### Expected Behavior

Concurrent calls for `Micronaut#start()` will launch multiple servers without errors.

### Actual Behaviour

The code in `Steps To Reproduce` part sometimes fails with the output into stderr:

```
io.micronaut.context.exceptions.DependencyInjectionException: Failed to inject value for parameter [enabled] of method [setEnabled] of class: io.micronaut.http.server.HttpServerConfiguration$CorsConfiguration

Message: Failed to inject value for parameter [enabled] of method [setEnabled] of class: io.micronaut.http.server.HttpServerConfiguration$CorsConfiguration

Message: Error resolving property value [micronaut.server.cors.enabled]. Property doesn't exist
Path Taken: NettyEmbeddedServer.buildDefaultServer(NettyHttpServerConfiguration configuration) --> new DefaultNettyEmbeddedServerFactory(ApplicationContext applicationContext,[RouteExecutor routeExecutor],MediaTypeCodecRegistry mediaTypeCodecRegistry,StaticResourceResolver staticResourceResolver,ThreadFactory nettyThreadFactory,HttpCompressionStrategy httpCompressionStrategy,EventLoopGroupFactory eventLoopGroupFactory,EventLoopGroupRegistry eventLoopGroupRegistry) --> new RouteExecutor([Router router],BeanContext beanContext,RequestArgumentSatisfier requestArgumentSatisfier,HttpServerConfiguration serverConfiguration,ErrorResponseProcessor errorResponseProcessor,ExecutorSelector executorSelector) --> new DefaultRouter([Collection builders]) --> new AnnotatedFilterRouteBuilder(BeanContext beanContext,ExecutionHandleLocator executionHandleLocator,UriNamingStrategy uriNamingStrategy,ConversionService conversionService,[ServerContextPathProvider contextPathProvider]) --> NettyHttpServerConfiguration.setCors([CorsConfiguration cors]) --> CorsConfiguration.setEnabled([boolean enabled])
Path Taken: NettyEmbeddedServer.buildDefaultServer(NettyHttpServerConfiguration configuration) --> new DefaultNettyEmbeddedServerFactory(ApplicationContext applicationContext,[RouteExecutor routeExecutor],MediaTypeCodecRegistry mediaTypeCodecRegistry,StaticResourceResolver staticResourceResolver,ThreadFactory nettyThreadFactory,HttpCompressionStrategy httpCompressionStrategy,EventLoopGroupFactory eventLoopGroupFactory,EventLoopGroupRegistry eventLoopGroupRegistry) --> new RouteExecutor([Router router],BeanContext beanContext,RequestArgumentSatisfier requestArgumentSatisfier,HttpServerConfiguration serverConfiguration,ErrorResponseProcessor errorResponseProcessor,ExecutorSelector executorSelector) --> new DefaultRouter([Collection builders]) --> new AnnotatedFilterRouteBuilder(BeanContext beanContext,ExecutionHandleLocator executionHandleLocator,UriNamingStrategy uriNamingStrategy,ConversionService conversionService,[ServerContextPathProvider contextPathProvider]) --> NettyHttpServerConfiguration.setCors([CorsConfiguration cors]) --> CorsConfiguration.setEnabled([boolean enabled])
	at io.micronaut.context.exceptions.DependencyInjectionException.missingProperty(DependencyInjectionException.java:289)
	at io.micronaut.context.AbstractInitializableBeanDefinition.lambda$resolvePropertyValue$10(AbstractInitializableBeanDefinition.java:2043)
	at java.base/java.util.Optional.orElseThrow(Optional.java:408)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolvePropertyValue(AbstractInitializableBeanDefinition.java:2043)
	at io.micronaut.context.AbstractInitializableBeanDefinition.getPropertyValueForSetter(AbstractInitializableBeanDefinition.java:1086)
	at io.micronaut.http.server.$HttpServerConfiguration$CorsConfiguration$Definition.injectBean(Unknown Source)
	at io.micronaut.http.server.$HttpServerConfiguration$CorsConfiguration$Definition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2354)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2305)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2251)
	at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:3016)
	at io.micronaut.context.SingletonScope.getOrCreate(SingletonScope.java:80)
	at io.micronaut.context.DefaultBeanContext.findOrCreateSingletonBeanRegistration(DefaultBeanContext.java:2918)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2879)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2800)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1617)
	at io.micronaut.context.AbstractBeanResolutionContext.getBean(AbstractBeanResolutionContext.java:66)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:2065)
	at io.micronaut.context.AbstractInitializableBeanDefinition.getBeanForSetter(AbstractInitializableBeanDefinition.java:1200)
	at io.micronaut.http.server.netty.configuration.$NettyHttpServerConfiguration$Definition.injectBean(Unknown Source)
	at io.micronaut.http.server.netty.configuration.$NettyHttpServerConfiguration$Definition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2354)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2305)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2251)
	at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:3016)
	at io.micronaut.context.SingletonScope.getOrCreate(SingletonScope.java:80)
	at io.micronaut.context.DefaultBeanContext.findOrCreateSingletonBeanRegistration(DefaultBeanContext.java:2918)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2879)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2800)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1617)
	at io.micronaut.context.AbstractBeanResolutionContext.getBean(AbstractBeanResolutionContext.java:66)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:2065)
	at io.micronaut.context.AbstractInitializableBeanDefinition.getBeanForConstructorArgument(AbstractInitializableBeanDefinition.java:1297)
	at io.micronaut.web.router.$AnnotatedFilterRouteBuilder$Definition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2354)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2305)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2251)
	at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:3016)
	at io.micronaut.context.SingletonScope.getOrCreate(SingletonScope.java:80)
	at io.micronaut.context.DefaultBeanContext.findOrCreateSingletonBeanRegistration(DefaultBeanContext.java:2918)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2879)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2853)
	at io.micronaut.context.DefaultBeanContext.addCandidateToList(DefaultBeanContext.java:3511)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistrations(DefaultBeanContext.java:3457)
	at io.micronaut.context.DefaultBeanContext.getBeanRegistrations(DefaultBeanContext.java:3427)
	at io.micronaut.context.DefaultBeanContext.getBeansOfType(DefaultBeanContext.java:1381)
	at io.micronaut.context.AbstractBeanResolutionContext.getBeansOfType(AbstractBeanResolutionContext.java:72)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBeansOfType(AbstractInitializableBeanDefinition.java:2161)
	at io.micronaut.context.AbstractInitializableBeanDefinition.getBeansOfTypeForConstructorArgument(AbstractInitializableBeanDefinition.java:1437)
	at io.micronaut.web.router.$DefaultRouter$Definition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2354)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2305)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2251)
	at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:3016)
	at io.micronaut.context.SingletonScope.getOrCreate(SingletonScope.java:80)
	at io.micronaut.context.DefaultBeanContext.findOrCreateSingletonBeanRegistration(DefaultBeanContext.java:2918)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2879)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2800)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1617)
	at io.micronaut.context.AbstractBeanResolutionContext.getBean(AbstractBeanResolutionContext.java:66)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:2065)
	at io.micronaut.context.AbstractInitializableBeanDefinition.getBeanForConstructorArgument(AbstractInitializableBeanDefinition.java:1297)
	at io.micronaut.http.server.$RouteExecutor$Definition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2354)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2305)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2251)
	at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:3016)
	at io.micronaut.context.SingletonScope.getOrCreate(SingletonScope.java:80)
	at io.micronaut.context.DefaultBeanContext.findOrCreateSingletonBeanRegistration(DefaultBeanContext.java:2918)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2879)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2800)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1617)
	at io.micronaut.context.AbstractBeanResolutionContext.getBean(AbstractBeanResolutionContext.java:66)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:2065)
	at io.micronaut.context.AbstractInitializableBeanDefinition.getBeanForConstructorArgument(AbstractInitializableBeanDefinition.java:1297)
	at io.micronaut.http.server.netty.$DefaultNettyEmbeddedServerFactory$Definition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2354)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2305)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2251)
	at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:3016)
	at io.micronaut.context.SingletonScope.getOrCreate(SingletonScope.java:80)
	at io.micronaut.context.DefaultBeanContext.findOrCreateSingletonBeanRegistration(DefaultBeanContext.java:2918)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2879)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2800)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1617)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1599)
	at io.micronaut.http.server.netty.$DefaultNettyEmbeddedServerFactory$BuildDefaultServer0$Definition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2354)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2305)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2251)
	at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:3016)
	at io.micronaut.context.SingletonScope.getOrCreate(SingletonScope.java:80)
	at io.micronaut.context.DefaultBeanContext.findOrCreateSingletonBeanRegistration(DefaultBeanContext.java:2918)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2879)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2800)
	at io.micronaut.context.DefaultBeanContext.findBean(DefaultBeanContext.java:1680)
	at io.micronaut.context.DefaultBeanContext.findBean(DefaultBeanContext.java:1655)
	at io.micronaut.context.DefaultBeanContext.findBean(DefaultBeanContext.java:878)
	at io.micronaut.context.BeanLocator.findBean(BeanLocator.java:291)
	at io.micronaut.runtime.Micronaut.start(Micronaut.java:77)
```

### Steps To Reproduce

```java
 static Micronaut buildMicronautContext(int portCandidate) {
        Micronaut micronaut = Micronaut.build("");

        Map<String, Object> properties = new HashMap<>();
        properties.putAll(serverProperties(portCandidate));

        return micronaut
                .properties(properties)
                .banner(false)
                .mapError(DependencyInjectionException.class, ex -> { // This is the actual exception that is thrown
                    System.out.println("Got exception: " + ex);
                    ex.printStackTrace();
                    return 111;
                });
    }

    static Map<String, Object> serverProperties(int port) {
        Map<String, Object> result = new HashMap<>();
        result.put("micronaut.server.port", port);
        result.put("micronaut.server.cors.enabled", "true");
        result.put("micronaut.server.cors.configurations.web.allowed-headers", "Authorization");

        return result;
    }

    public static void main(String[] args) throws ExecutionException, InterruptedException {
        AtomicInteger portShift = new AtomicInteger(0);
        ExecutorService executorService = Executors.newFixedThreadPool(16);
        CopyOnWriteArrayList<Future<ApplicationContext>> futures = new CopyOnWriteArrayList<>();

        for (int i = 0; i < 50; i++) {
            var f = executorService.submit(() -> {
                try {
                    int port = 10300 + portShift.getAndIncrement();
                    var micronaut = buildMicronautContext(port);
                    System.out.println("Starting server on: " + port);
                    return micronaut.start();
                } catch (Throwable th) {
                    System.out.println("Failed to start: " + th.getMessage());
                    return null;
                }
            });
            futures.add(f);
        }

        ArrayList<ApplicationContext> contexts = new ArrayList<>();
        for (Future<ApplicationContext> future : futures) {
            try {
                ApplicationContext ctx = future.get();
                contexts.add(ctx);
                System.out.println("Started on: " + ctx.getEnvironment().getProperty("micronaut.server.port", Integer.class).orElse(-1));
            } catch (Throwable th) {
                System.out.println("Failed to start 2: " + th.getMessage());
            }
        }

        for (ApplicationContext context : contexts) {
            try {
                context.stop();
            } catch (Throwable th) {
                System.out.println("Failed to stop: " + th.getMessage());
            }
        }
        executorService.shutdown();
    }
```

### Environment Information

- JDK version `java version "17.0.8" 2023-07-18 LTS`

### Example Application

_No response_

### Version

3.10.x

### PR DESCRIPTION
**Pull request by PakhomovAlexander**
The usage of `typeConverters` and `converterCache` are optimized. 

1) Redundant operations are removed, which boosts the performance (according to the  `ConversionServiceBenchmark`) at ~ 2x.

2) The synchronized keyword is added to the critical methods that can be called from several threads. Now `typeConverters` and `converterCache` are consistent.

Here are the benchmark results that I've run on my Macbook 2,6 GHz 6-Core Intel Core i7, jdk-11.0.12.jdk:

```
Benchmark                                     Mode  Cnt         Score          Error  Units

Baseline:
ConversionServiceBenchmark.convertCacheHit   thrpt    5  27404123.384 ± 5692431.522  ops/s
ConversionServiceBenchmark.convertCacheMiss  thrpt    5   4923550.293 ±  461767.609  ops/s

Refactored (1):
ConversionServiceBenchmark.convertCacheHit   thrpt    5  48408342.853 ± 9369123.821  ops/s
ConversionServiceBenchmark.convertCacheMiss  thrpt    5   4842578.352 ±   34813.903  ops/s

Refactored + synchronized (1 + 2):
ConversionServiceBenchmark.convertCacheHit   thrpt    5  38918162.398 ± 10242478.879  ops/s
ConversionServiceBenchmark.convertCacheMiss  thrpt    5   4512350.735 ±   981537.261  ops/s
```

### CODE REVIEW

**Comment by PakhomovAlexander** [comment]
Also, the original issue is not reproduced in this fix. 

@wetted Do you think this fix is good enough?

**Comment by wetted** [comment]
@PakhomovAlexander thank you for the PR. I'm asking for reviews from those who know micronaut core better than I.

**Comment by sdelamo** [comment]
in which scenario did you get the race condition. In tests?

**Comment by PakhomovAlexander** [comment]
@sdelamo I've described the case with the race condition in this issue https://github.com/micronaut-projects/micronaut-core/issues/10091 

Here is the repo wich reproduces it https://github.com/wetted/core-10091-race-condition 

TLDR; Yes, first I got the race condition on tests but then I managed to reproduce it in the main code. Basically, any application that starts context from different threads can face this issue.

**Review by dstepanov — APPROVED** [review]

**Review by yawkat — APPROVED** [review]

**Comment by PakhomovAlexander** [comment]
Hi @sdelamo, could you please clarify the state of this PR? I wonder if the fix will appear in the upcoming 3.10.x release. Thanks!

**Comment by sdelamo** [comment]
@PakhomovAlexander it will be part of a 3.10.x release.
